File: app/src/main/java/com/shadow/eye/ShadowEyeActivity.java
package com.shadow.eye;

import android.app.Activity;
import android.content.Context;
import android.graphics.SurfaceTexture;
import android.hardware.Camera;
import android.hardware.Camera.AutoFocusCallback;
import android.hardware.Camera.Parameters;
import android.hardware.Camera.PictureCallback;
import android.hardware.Camera.PreviewCallback;
import android.media.AudioManager;
import android.os.Bundle;
import android.os.Environment;
import android.os.PowerManager;
import android.os.PowerManager.WakeLock;
import android.os.SystemClock;
import android.os.Vibrator;
import android.view.KeyEvent;
import android.view.Window;
import android.view.WindowManager;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

public class ShadowEyeActivity extends Activity {
    private static final int NULL_MODE = 0, PHOTO_MODE = 1;
    private static final int CAMERA_PREVIEW_WIDTH = 500, CAMERA_PREVIEW_HEIGHT = 500;
    private final SurfaceTexture mSurfaceTexture = new SurfaceTexture(0);
    private WakeLock mWakeLock = null;
    private AudioManager mAudioManager;
    private Camera mCamera = null;
    private final AutoFocusCallback mAutoFocusCallback = new AutoFocusCallback() {
        @Override
        public void onAutoFocus(boolean success, Camera camera) {
            if (success) {
                try {
                    camera.takePicture(null, null, mPictureCallback);
                } catch (Exception e) {
                    closeCamera();
                }
            } else {
                mCamera.autoFocus(mAutoFocusCallback);
            }
        }
    };
    private Vibrator mVibrator = null;
    private String mCurrentPath;
    private long mLastBackPress = 0;
    private int mCameraStatue = NULL_MODE;
    private boolean mBackCamera = true;
    private final PreviewCallback mPreviewCallback = new PreviewCallback() {
        final int autoFocusLightThreshold = 16;
        double lastLight = 100000;
        long stableTime = 0;
        boolean isFocused = false;

        public void onPreviewFrame(byte[] data, Camera camera) {
            try {
                double light = getYUV420SPLight(data, CAMERA_PREVIEW_WIDTH, CAMERA_PREVIEW_HEIGHT);
                long currentTime = SystemClock.elapsedRealtime();
                // Light Change
                if (Math.abs(lastLight - light) > autoFocusLightThreshold) {
                    lastLight = light;
                    stableTime = currentTime;
                    isFocused = false;
                } else if (!isFocused) { // If Change Little ,Focus Camera
                    if (currentTime - stableTime > 500) {
                        mVibrator.vibrate(50);
                        mCamera.setPreviewCallback(null);
                        if (mBackCamera) {
                            camera.autoFocus(mAutoFocusCallback);
                        } else {
                            try {
                                camera.takePicture(null, null, mPictureCallback);
                            } catch (Exception e) {
                                closeCamera();
                            }
                        }
                        isFocused = true;
                    }
                }

            } catch (Exception e) {
                closeCamera();
            }
        }
    };
    private final PictureCallback mPictureCallback = new PictureCallback() {
        @Override
        public void onPictureTaken(byte[] data, Camera camera) {
            long l[] = {0, 100, 100, 100};
            mVibrator.vibrate(l, -1);
            savePicture(data);
            mCamera.startPreview();
            mCamera.setPreviewCallback(mPreviewCallback);
        }
    };
    private long mLastVolumeKeyPressTime = 0;

    /**
     * 通过YUV420SP数组 计算出中心区域亮度
     *
     * @param yuv420sp 图像帧数组
     * @param width    宽度
     * @param height   高度
     * @return 返回亮度, 这亮度值始终大于0
     */
    public static double getYUV420SPLight(byte[] yuv420sp, int width,
                                          int height) {
        double bright = 0;
        int total = 0;
        final int xEdge = width / 3, yEdge = height / 3, focusWidth = width - xEdge * 2, focusHeight = height - yEdge * 2;
        final int dt = 32, dx = (focusWidth <= dt ? 1 : focusWidth / dt), dy = (focusHeight <= dt ? 1 : focusHeight / dt);
        for (int j = yEdge; j < height - yEdge; j += dy) {
            for (int i = xEdge; i < width - xEdge; i += dx, ++total) {
                bright += (0xff & ((int) yuv420sp[j * width + i]));
            }
        }
        return 0 == total ? 0 : (bright / total);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        mAudioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
        mVibrator = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);
        setContentView(R.layout.activity_main);
        setBrightness(1);
        acquireWakeLock();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        releaseWakeLock();
        closeCamera();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        // 获取手机当前音量值
        switch (keyCode) {
            case KeyEvent.KEYCODE_VOLUME_DOWN:
                handleKeyPress(false);
                mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC,
                        AudioManager.ADJUST_LOWER, 0);
                mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC,
                        AudioManager.ADJUST_RAISE, 0);
                return true;

            case KeyEvent.KEYCODE_VOLUME_UP:
                handleKeyPress(true);
                mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC,
                        AudioManager.ADJUST_LOWER, 0);
                mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC,
                        AudioManager.ADJUST_RAISE, 0);
                return true;
            case KeyEvent.KEYCODE_HOME:
                return true;
            default:
                break;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public void onBackPressed() {
        long currentTime = SystemClock.elapsedRealtime();
        if (currentTime - mLastBackPress > 500) {
            mLastBackPress = currentTime;
        } else {
            super.onBackPressed();
        }
    }

    private void savePicture(byte[] data) {
        if (mCurrentPath == null) return;
        String dateString = getDateString();
        String Path = mCurrentPath + "/" + dateString + ".zip";
        new File(mCurrentPath).mkdirs();
        ZipOutputStream zipOutputStream;
        try {
            zipOutputStream = new ZipOutputStream(new FileOutputStream(new File(Path)));
            try {
                ZipEntry zipEntry = new ZipEntry(dateString + ".jpg");
                zipOutputStream.putNextEntry(zipEntry);
                zipOutputStream.write(data);
                zipOutputStream.flush();
                zipOutputStream.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        } catch (Exception e1) {
            e1.printStackTrace();
        }
    }

    private String getDateString() {
        return new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss", Locale.CHINESE).format(new Date());
    }

    private void setBrightness(float brightness) {
        WindowManager.LayoutParams params = getWindow().getAttributes();
        params.flags |= WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON;
        params.screenBrightness = brightness / 255.0f;
        getWindow().setAttributes(params);
    }

    private void handleKeyPress(boolean backCamera) {
        mVibrator.vibrate(50);
        if (mCameraStatue == NULL_MODE) {
            long currentTime = SystemClock.elapsedRealtime();
            if (mLastVolumeKeyPressTime + 3000 < currentTime) {
                mLastVolumeKeyPressTime = currentTime;
                mBackCamera = backCamera;
                return;
            }
            mLastVolumeKeyPressTime = 0;
            if (mCamera == null) {
                startCamera();
            }
        }
        if (mBackCamera == backCamera) {
            if (mCameraStatue == PHOTO_MODE) {
                mCamera.takePicture(null, null, mPictureCallback);
            } else if (mCameraStatue == NULL_MODE) {
                mCameraStatue = PHOTO_MODE;
                mCamera.setPreviewCallback(mPreviewCallback);
            }
        } else {
            closeCamera();
            mCameraStatue = NULL_MODE;
        }
    }

    private void closeCamera() {
        if (mCamera != null) {
            mCamera.setPreviewCallback(null);
            mCamera.stopPreview();
            mCamera.release();
            mCamera = null;
        }
    }

    private int findCamera(boolean back) {
        Camera.CameraInfo cameraInfo = new Camera.CameraInfo();
        int cameraCount = Camera.getNumberOfCameras(); // get cameras number
        int value = back ? Camera.CameraInfo.CAMERA_FACING_BACK : Camera.CameraInfo.CAMERA_FACING_FRONT;
        for (int camIdx = 0; camIdx < cameraCount; camIdx++) {
            Camera.getCameraInfo(camIdx, cameraInfo);
            if (cameraInfo.facing == value) {
                return camIdx;
            }
        }
        return -1;
    }

    // Main Func of Catch Camera
    private Camera openCamera(boolean backCamera) {
        int cameraIndex = findCamera(backCamera);
        if (cameraIndex == -1) {
            return null;
        }
        try {
            Camera camera = Camera.open(cameraIndex);
            camera.setDisplayOrientation(90);

            try {
                Parameters p = camera.getParameters();
                if (p != null) {
                    if (backCamera) {
                        p.setFlashMode(Parameters.FLASH_MODE_OFF);
                        p.set("orientation", "portrait");
                        p.setRotation(90);
                    } else {
                        p.set("orientation", "portrait");
                        p.setRotation(270);
                    }
                    p.setPreviewSize(CAMERA_PREVIEW_WIDTH, CAMERA_PREVIEW_HEIGHT);
                    p.setAntibanding(Parameters.ANTIBANDING_60HZ);
                    camera.setParameters(p);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            camera.setPreviewTexture(mSurfaceTexture);
            return camera;
        } catch (Exception e1) {
            return null;
        }
    }

    private void startCamera() {
        mCamera = openCamera(mBackCamera);
        if (mCamera != null) {
            mCurrentPath = Environment.getExternalStorageDirectory() + "/save/" + getDateString();
            mCamera.startPreview();
        }
    }

    private void acquireWakeLock() {
        if (null == mWakeLock) {
            PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
            mWakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK
                    | PowerManager.ON_AFTER_RELEASE, "LOCK");
            if (null != mWakeLock) {
                mWakeLock.acquire();
            }
        }
    }

    // 释放设备电源锁
    private void releaseWakeLock() {
        if (null != mWakeLock) {
            mWakeLock.release();
            mWakeLock = null;
        }
    }
}

File: app/src/main/java/com/shadow/eye/ShadowEyeActivity.kt
package com.shadow.eye

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.SurfaceTexture
import android.hardware.Camera
import android.hardware.Camera.*
import android.media.AudioManager
import android.os.*
import android.os.PowerManager.WakeLock
import android.view.KeyEvent
import android.view.View
import android.view.Window
import android.view.WindowManager
import kotlinx.android.synthetic.main.activity_main.*
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.hardware.Camera.Parameters.FOCUS_MODE_AUTO
import java.io.ByteArrayOutputStream


class ShadowEyeActivity : Activity() {
    private val mSurfaceTexture = SurfaceTexture(0)
    private var mWakeLock: WakeLock? = null
    private var mAudioManager: AudioManager? = null
    private var mCamera: Camera? = null
    private var mCameraInfo: CameraInfo? = null
    private val mAutoFocusCallback: AutoFocusCallback = object : AutoFocusCallback {
        override fun onAutoFocus(success: Boolean, camera: Camera) {
            if (success) {
                try {
                    camera.takePicture(null, null, mPictureCallback)
                } catch (e: Exception) {
                    closeCamera()
                }

            } else {
                mCamera?.autoFocus(this)
            }
        }
    }
    private var mVibrator: Vibrator? = null
    private var mCurrentPath: String? = null
    private var mLastBackPress: Long = 0
    private var mCameraStatue = NULL_MODE
    private var mBackCamera = true
    private val mPreviewCallback: PreviewCallback = object : PreviewCallback {
        val autoFocusLightThreshold = 16
        var lastLight = 100000.0
        var stableTime: Long = 0
        var isFocused = false

        override fun onPreviewFrame(data: ByteArray, camera: Camera) {
            try {
                val previewSize = camera.parameters.previewSize
                val light = getYUV420SPLight(data, previewSize.width, previewSize.height)
                val currentTime = SystemClock.elapsedRealtime()
                // Light Change
                if (Math.abs(lastLight - light) > autoFocusLightThreshold) {
                    lastLight = light
                    stableTime = currentTime
                    isFocused = false
                } else if (!isFocused) { // If Change Little, Focus Camera
                    if (currentTime - stableTime > 500) {
                        mVibrator?.vibrate(50)
                        mCamera?.setPreviewCallback(null)
                        if (mBackCamera) {
                            camera.autoFocus(mAutoFocusCallback)
                        } else {
                            try {
                                camera.takePicture(null, null, mPictureCallback)
                            } catch (e: Exception) {
                                closeCamera()
                            }

                        }
                        isFocused = true
                    }
                }

            } catch (e: Exception) {
                closeCamera()
            }

        }
    }
    private val mPictureCallback = PictureCallback { data, camera ->
        mVibrator?.vibrate(longArrayOf(0, 100, 100, 100), -1)
        savePicture(rotateJpeg(data, mCameraInfo!!.orientation))
        mCamera?.startPreview()
        mCamera?.setPreviewCallback(mPreviewCallback)
    }

    private fun rotateJpeg(data: ByteArray, orientation: Int): ByteArray {
        if (orientation == 0) return data
        val bitmap = BitmapFactory.decodeByteArray(data, 0, data.size)
        val newBitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height,
                Matrix().apply { setRotate(orientation.toFloat(), bitmap.width.toFloat() / 2, bitmap.height.toFloat() / 2) }, true)
        return ByteArrayOutputStream().apply { newBitmap.compress(Bitmap.CompressFormat.JPEG, 100, this) }.toByteArray()
    }

    private var mLastVolumeKeyPressTime = 0L

    private val dateString: String
        get() = SimpleDateFormat("yyyy-MM-dd_HH-mm-ss", Locale.CHINESE).format(Date())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            window.attributes.systemUiVisibility = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or View.SYSTEM_UI_FLAG_IMMERSIVE
        }
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
        mAudioManager = getSystemService(Context.AUDIO_SERVICE) as AudioManager
        mVibrator = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        setContentView(R.layout.activity_main)
        setBrightness(1f)
        acquireWakeLock()

        viewFinish.setOnClickListener { onBackPressed() } // top -> finish
        viewBack.setOnClickListener { handleKeyPress(true) } // middle -> back
        viewFront.setOnClickListener { handleKeyPress(false) } // bottom -> front
    }

    override fun onDestroy() {
        super.onDestroy()
        releaseWakeLock()
        closeCamera()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        // 获取手机当前音量值
        when (keyCode) {
            KeyEvent.KEYCODE_VOLUME_DOWN -> {
                handleKeyPress(false)
                mAudioManager!!.adjustStreamVolume(AudioManager.STREAM_MUSIC,
                        AudioManager.ADJUST_LOWER, 0)
                mAudioManager!!.adjustStreamVolume(AudioManager.STREAM_MUSIC,
                        AudioManager.ADJUST_RAISE, 0)
                return true
            }

            KeyEvent.KEYCODE_VOLUME_UP -> {
                handleKeyPress(true)
                mAudioManager!!.adjustStreamVolume(AudioManager.STREAM_MUSIC,
                        AudioManager.ADJUST_LOWER, 0)
                mAudioManager!!.adjustStreamVolume(AudioManager.STREAM_MUSIC,
                        AudioManager.ADJUST_RAISE, 0)
                return true
            }
            KeyEvent.KEYCODE_HOME -> return true
            else -> {
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onBackPressed() {
        val currentTime = SystemClock.elapsedRealtime()
        if (currentTime - mLastBackPress > 500) {
            mLastBackPress = currentTime
        } else {
            super.onBackPressed()
        }
    }

    private fun savePicture(data: ByteArray) {
        if (mCurrentPath == null) return
        val dateString = dateString
        val Path = "$mCurrentPath/$dateString.zip"
        File(mCurrentPath!!).mkdirs()
        var zipOutputStream: ZipOutputStream? = null
        try {
            zipOutputStream = ZipOutputStream(FileOutputStream(File(Path)))
            zipOutputStream.putNextEntry(ZipEntry("$dateString.jpg"))
            zipOutputStream.write(data)
            zipOutputStream.flush()
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            zipOutputStream?.close()
        }

    }

    private fun setBrightness(brightness: Float) {
        val params = window.attributes
        params.flags = params.flags or WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
        params.screenBrightness = brightness / 255.0f
        window.attributes = params
    }

    private fun handleKeyPress(backCamera: Boolean) {
        mVibrator!!.vibrate(50)
        if (mCameraStatue == NULL_MODE) {
            val currentTime = SystemClock.elapsedRealtime()
            if (mLastVolumeKeyPressTime + 3000 < currentTime) {
                mLastVolumeKeyPressTime = currentTime
                mBackCamera = backCamera
                return
            }
            mLastVolumeKeyPressTime = 0
            if (mCamera == null) {
                startCamera()
            }
        }
        if (mBackCamera == backCamera) {
            if (mCameraStatue == PHOTO_MODE) {
                mCamera?.takePicture(null, null, mPictureCallback)
            } else if (mCameraStatue == NULL_MODE) {
                mCameraStatue = PHOTO_MODE
                mCamera?.setPreviewCallback(mPreviewCallback)
            }
        } else {
            closeCamera()
            mCameraStatue = NULL_MODE
        }
    }

    private fun closeCamera() {
        if (mCamera != null) {
            mCamera?.setPreviewCallback(null)
            mCamera?.stopPreview()
            mCamera?.release()
            mCamera = null
            mCameraInfo = null
        }
    }

    private fun startCamera() {
        var cameraIndex = -1
        val cameraInfo = Camera.CameraInfo()
        val cameraFacing = if (mBackCamera) Camera.CameraInfo.CAMERA_FACING_BACK else Camera.CameraInfo.CAMERA_FACING_FRONT
        for (camIdx in 0 until Camera.getNumberOfCameras()) {
            Camera.getCameraInfo(camIdx, cameraInfo)
            if (cameraInfo.facing == cameraFacing) {
                cameraIndex = camIdx
                break
            }
        }
        if (cameraIndex == -1) {
            return
        }

        try {
            val camera = Camera.open(cameraIndex)
            camera.setDisplayOrientation(cameraInfo.orientation)

            try {
                val parameters = camera.parameters
                if (parameters != null) {
                    if (mBackCamera) {
                        parameters.flashMode = Parameters.FLASH_MODE_OFF
                    }
                    parameters.setRotation(cameraInfo.orientation)
                    parameters.sceneMode = Parameters.SCENE_MODE_PORTRAIT
                    val pictureSize = parameters.supportedPictureSizes.firstOrNull()
                    if (pictureSize != null) parameters.setPictureSize(pictureSize.width, pictureSize.height)
                    parameters.antibanding = Parameters.ANTIBANDING_60HZ
                    parameters.focusMode = FOCUS_MODE_AUTO
                    camera.parameters = parameters
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }

            camera.setPreviewTexture(mSurfaceTexture)
            mCurrentPath = Environment.getExternalStorageDirectory().toString() + "/save/" + dateString
            camera.startPreview()
            mCamera = camera
            mCameraInfo = cameraInfo
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @SuppressLint("InvalidWakeLockTag", "WakelockTimeout")
    private fun acquireWakeLock() {
        if (null == mWakeLock) {
            val pm = getSystemService(Context.POWER_SERVICE) as PowerManager
            mWakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK or PowerManager.ON_AFTER_RELEASE, "LOCK")
            mWakeLock?.acquire()
        }
    }

    // 释放设备电源锁
    private fun releaseWakeLock() {
        if (null != mWakeLock) {
            mWakeLock!!.release()
            mWakeLock = null
        }
    }

    companion object {
        private val NULL_MODE = 0
        private val PHOTO_MODE = 1

        /**
         * 通过YUV420SP数组 计算出中心区域亮度
         *
         * @param yuv420sp 图像帧数组
         * @param width    宽度
         * @param height   高度
         * @return 返回亮度, 这亮度值始终大于0
         */
        fun getYUV420SPLight(yuv420sp: ByteArray, width: Int,
                             height: Int): Double {
            var bright = 0.0
            var total = 0
            val xEdge = width / 3
            val yEdge = height / 3
            val focusWidth = width - xEdge * 2
            val focusHeight = height - yEdge * 2
            val dt = 32
            val dx = if (focusWidth <= dt) 1 else focusWidth / dt
            val dy = if (focusHeight <= dt) 1 else focusHeight / dt
            var j = yEdge
            while (j < height - yEdge) {
                var i = xEdge
                while (i < width - xEdge) {
                    bright += (0xff and yuv420sp[j * width + i].toInt()).toDouble()
                    i += dx
                    ++total
                }
                j += dy
            }
            return if (0 == total) 0.0 else bright / total
        }
    }
}